src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
+                        [[
+                            sql_param(&LTree::from(traversal_path.clone())),
+                            sql_param(&created),
+                            sql_param(&hash_id),
+                            sql_param(&children_values),
+                            sql_param(&children_bitvec),
+                        ]],
+                    )
+                    .await?;
+                }
+
+                // We only care about the Branch Nodes here
+                // Leaf node is already inserted
+                _ => continue,
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<Types: NodeType> MerklizedStateDataSource<Types> for SqlStorage {
+    type Error = QueryError;
+
+    /// Retreives a Merkle path from the database
+    async fn get_path<
+        E: Element + Send + DeserializeOwned,
+        I: Index + Send + ToTraversalPath<A> + DeserializeOwned,
+        A: Unsigned,
+        T: NodeValue + Send + CanonicalDeserialize,
+    >(
+        &self,
+        state_type: &'static str,
+        tree_height: usize,
+        header_state_commitment_field: &'static str,
+        snapshot: Snapshot<Types>,
+        key: serde_json::Value,
+    ) -> QueryResult<MerklePath<E, I, T>> {
+        let index: I = serde_json::from_value(key).map_err(ParseError::Serde)?;
+        // Get the traversal path of the index
+        let path = I::to_traversal_path(&index, tree_height)
+            .into_iter()
+            .map(|x| x as i64)
+            .collect::<Vec<_>>();
+
+        let created = match snapshot {
+            Snapshot::<Types>::Commit(commit) => {
+                let commit: &[u8; 32] = commit.as_ref();
+                // Get the block height using the merkle commitment.
+                let query = self
+                    .query_one(
+                        &format!(
+                            "SELECT n.created FROM {state_type} n INNER JOIN Header h on 
+                            n.created = h.height where h.data->>{header_state_commitment_field}"
+                        ),
+                        &[&commit.to_vec()],
+                    )
+                    .await?;
+
+                query.get(0)
+            }
+            Snapshot::<Types>::Index(created) => {
+                let created = created as i64;
+                // Check if header exists
+                // For Snapshot::Commit the entry is already checked when retrieving the block height
+                self.query_one(
+                    "SELECT 1 from HEADER where height = $1",
+                    [sql_param(&created)],
+                )
+                .await?;
+
+                created
+            }
+        };
+
+        //  Get all the nodes in the path to the index.
+        // Order by pos DESC is to return nodes from the leaf to the root
+        let nodes = self
+            .query(
+                &format!(
+                    " WITH recent_nodes AS (
```

Comment:
What's the reason for not using `SELECT DISTINCT ON` to get the most recent version of each node in the path? IMO that is easier to read than this subquery/self-join

Reply:
I think distinct on just selects a random node and the results are then results are ordered by which is wrong.

Reply:
I think distinct on respects the order by clause. From [the docs](https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT):

> Note that the “first row” of each set is unpredictable **unless ORDER BY is used to ensure that the desired row appears first**

They given an example of doing pretty much exactly what we're trying to do

Reply:
aah okay, one of the test was failing because of query but I think that was related to something else. I'll change it back